### DISCUSSION TITLE
Possible to define a default/fallback config?

### DISCUSSION THREAD

**Discussion by BryceBeagle** [astral-sh/ruff#12063] [Q&A]
Hey all,

We're working on implementing a ruff framework in our monorepo where we have the following constraints
1. We have a central config file. Let's call it `/config/ruff.toml`
2. We have a bunch of python libraries scattered around our codebase. 
    - It's important to note that the central config is not in a directory directly "above" the libraries, so the normal relative "[config file discovery](https://docs.astral.sh/ruff/configuration/#config-file-discovery)" doesn't apply
3. Some of these libraries will have their own `ruff.toml`(s)
    - They have `extend = "/config/ruff.toml"`, and then either enable or disable a couple specific rules
4. Other libraries will not have their own `ruff.toml`s
5. Either way, `/config/ruff.toml`'s rules should be the default for _all code_, unless otherwise overridden in a local `ruff.toml`

Some approaches we considered:
* Always providing `--config=/config/ruff.toml`
    - this doesn't work because then the local config files are completely ignored
* Using multiple `--config` flags and passing both the central and local configs
    - this doesn't work because [using multiple `--config` flags is not supported](https://github.com/astral-sh/ruff/issues/10035#issuecomment-2002135867)
* Never providing `--config` with a file and instead using `--config "extend = '/config/ruff.toml'"`
    - this doesn't work because `extend` [is explicitly blocked](https://github.com/astral-sh/ruff/issues/10035) from being used with `--config`
* Only providing `--config` if the the library doesn't have _any_ `ruff.toml` of it's own
    - this _kinda_ works but falls apart if _some_ subdirectories of the library have `ruff.toml`s, but not all.
    - If this is the case, some directories will not use _any_ config, which is undesirable
* Force-overriding `$XDG_CONFIG_HOME=/config` so that the default config fallback described [here](https://docs.astral.sh/ruff/configuration/#config-file-discovery:~:text=If%20no%20config%20file%20is%20found%20in%20the%20filesystem%20hierarchy%2C%20Ruff%20will%20fall%20back%20to%20using%20a%20default%20configuration) would pick it as the default
    - This is gross and could _potentially_ interfere with other things in our pipeline that rely on the env var/`dirs` crate
    - We also have developers on MacOS and hacking around `$HOME/Library/Application Support` too is even less tenable

The question is: is there a better way to do this without getting hacky?

**Comment by MichaReiser** [comment]
That's an interesting use case. I don't think there's currently a better way of doing what you want. How do other tools discover your configuration in `/config/`? Or is Ruff the only tool where you need this specific behavior?

**Reply by BryceBeagle** [reply]
> How do other tools discover your configuration in /config/? Or is Ruff the only tool where you need this specific behavior?

Admittedly, at the moment, `ruff` is the only tool with this requirement. We're implementing some generic tooling for linting in the monorepo with Bazel (starting with `ruff`), and the idea was to have the config files live with the tooling.
